v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -0,0 +1,455 @@
+                    # handler is also called. Why? Who knows.
+                    # But we don't wish to issue spurious notifications or event loop stoppages.
+                    logger.debug(
+                        "Connect Failure Disconnect Response: rc {} - {}".format(rc, message)
+                    )
+                else:
+                    # Double disconnect. Suppress.
+                    # Sometimes Paho disconnects twice. Why? Who knows.
+                    # But we don't wish to issue spurious notifications or event loop stoppages.
+                    logger.debug("Double Disconnect Response: rc {} - {}".format(rc, message))
+
+            if do_set_disconnect:
+                # Change state and notify tasks waiting on disconnect
+                async def set_disconnected():
+                    logger.debug("Client State: DISCONNECTED")
+                    this._connected = False
+                    async with this.disconnected_cond:
+                        this.disconnected_cond.notify_all()
+
+                f = asyncio.run_coroutine_threadsafe(set_disconnected(), this._loop)
+                # Need to wait for this one to finish since we don't want to let another
+                # Paho handler invoke until we know the connection state has been set.
+                f.result()
+
+            if do_stop_network_loop:
+
+                async def stop_network_loop():
+                    logger.debug("Stopping Paho network loop due to unexpected disconnect")
+                    client.loop_stop()
+
+                # NOTE: This coroutine can't finish execution until after this handler exits
+                # since this handler itself is running on the network loop.
+                # This means this MUST be the last task that gets scheduled in this handler,
+                # since .loop_stop() will block until this handler finishes running.
+                asyncio.run_coroutine_threadsafe(stop_network_loop(), this._loop)
+
+        def on_subscribe(client, userdata, mid):
+            pass
+
+        def on_unsubscribe(client, userdata, mid):
+            pass
+
+        mqtt_client.on_connect = on_connect
+        mqtt_client.on_disconnect = on_disconnect
+        mqtt_client.on_subscribe = on_subscribe
+        mqtt_client.on_unsubscribe = on_unsubscribe
+
+        return mqtt_client
+
+    async def _reconnect_loop(self):
+        logger.debug("Reconnect Daemon starting...")
+        current_attempt = None
+        try:
+            while True:
+                async with self.disconnected_cond:
+                    await self.disconnected_cond.wait_for(
+                        lambda: not self.is_connected() and self._desire_connection
+                    )
+                try:
+                    logger.debug("Reconnect Daemon attempting to reconnect...")
+                    current_attempt = asyncio.create_task(self.connect())
+                    await current_attempt
```

Comment:
Why not just do `await self.connect()`?

Reply:
So that I can cancel the connect task if the reconnect daemon task is cancelled. Not completely sure that it's necessary (need to investigate cancel more), but it seemed safer for now.

Reply:
Fair enough. I really hope that it's not necessary.

Reply:
Fixed